airflow/contrib/operators/gcp_function_operator.py

```diff
@@ -0,0 +1,571 @@
+            raise FieldValidationException("The key '{}' does not match regexp '{}'".
+                                           format(key, LABEL_KEY_REGEXP))
+        value = the_dictionary[key]
+        if not re.match(LABEL_VALUE_REGEXP, value):
+            raise FieldValidationException(
+                "The value '{}' of key '{}' does not match regexp '{}'".
+                format(value, key, LABEL_VALUE_REGEXP))
+
+
+ENV_VARS_KEY_REGEXP = r'^[^\x00=]+$'
+ENV_VARS_VALUE_REGEXP = r'^[^\\x00]*$'
+
+
+def _validate_env_vars_dict(the_dictionary):
+    # type: (dict) -> None
+    if not isinstance(the_dictionary, dict):
+        raise FieldValidationException("The field should be dictionary!")
+    for key in the_dictionary.keys():
+        if key is None:
+            raise FieldValidationException("The key '{}' cannot be None".format(key))
+        if not re.match(ENV_VARS_KEY_REGEXP, key):
+            raise FieldValidationException("The key '{}' does not match regexp '{}'".
+                                           format(key, ENV_VARS_KEY_REGEXP))
+        value = the_dictionary[key]
+        if not re.match(ENV_VARS_VALUE_REGEXP, value):
+            raise FieldValidationException(
+                "The value '{}' of key '{}' does not match regexp '{}'".
+                format(value, key, ENV_VARS_VALUE_REGEXP))
+
+
+CLOUD_FUNCTION_VALIDATION = [
+    dict(name="name", regexp=r'^[^\s]+$'),
+    dict(name="description", regexp="^.+$", optional=True),
+    dict(name="entryPoint", optional=True, regexp=r'^\w+$'),
+    dict(name="runtime", regexp=r'^nodejs6$|^nodejs8$|^python37$', optional=True),
+    dict(name="timeout", regexp=r'^\d+(\.\d{1,9})?s$', optional=True),
+    dict(name="availableMemoryMb", custom_validation=_validate_available_memory_in_mb,
+         optional=True),
+    dict(name="labels", custom_validation=_validate_labels_dict, optional=True),
+    dict(name="environmentVariables", custom_validation=_validate_env_vars_dict,
+         optional=True),
+    dict(name="network",
+         regexp=r'^[^\s]+$',
+         optional=True),
+    dict(name="maxInstances", optional=True, custom_validation=_validate_max_instances),
+
+    dict(name="source_code", type="union", fields=[
+        dict(name="sourceArchiveUrl", regexp=r'^gs://[^\s]+$'),
+        dict(name="sourceRepositoryUrl", regexp=r'^gs://[^\s]+$', api='v1beta2'),
+        dict(name="sourceRepository", type="dict", fields=[
+            dict(name="url",
+                 regexp=r'^[^\s]+$')
+        ]),
+        dict(name="sourceUploadUrl", regexp=r'^https://[^\s]+$')
+    ]),
+
+    dict(name="trigger", type="union", fields=[
+        dict(name="httpsTrigger", type="dict", fields=[
+            # This dict should be empty at input (url is added at output)
+        ]),
+        dict(name="eventTrigger", type="dict", fields=[
+            dict(name="eventType", regexp=r'^[^\s]+$'),
+            dict(name="resource", regexp=r'^[^/]+$'),
```

Comment:
Does this regexp matches 'projects/_/buckets/myBucket' for eventTrigger.resource in this doc?: https://cloud.google.com/functions/docs/reference/rest/v1/projects.locations.functions#EventTrigger

Reply:
Fixed in cb0bba61097c9c6a45538392015a4394f70e8905

Reply:
I think this should be solved consistently for all GCP operators, not just Cloud Functions.

I think there is a non-trivial value of early validation: (IIUC, it can catch e.g. syntax errors, like spurious whitespaces, before the workflow is even started - which may be hours before the operator is actually executed, right? If so, I think that there is a balance to strike here, and some regexps (e.g. enforcing documented constraints on resource names) are safe and useful.